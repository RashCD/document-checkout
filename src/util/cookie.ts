import Cookies from 'js-cookie';

class Cookie {
  static DEFAULT_OPTIONS = {
    path: '/',
    secure: window.location.protocol === 'https:',
    domain: window.location.hostname || '',
    expires: 365,
  };

  static set(key: string, value: string, options = {}) {
    return Cookies.set(key, value, {
      ...Cookie.DEFAULT_OPTIONS,
      ...options,
    });
  }

  static append<T>(key: string, value: T, options = {}) {
    const cookieValue = Cookie.get(key, '');

    if (cookieValue) {
      const cookieValueArray: T[] = JSON.parse(cookieValue);

      const cookieExist = cookieValueArray.some((cookie) => {
        return JSON.stringify(cookie) === JSON.stringify(value);
      });

      const mergeCookieValue = [...cookieValueArray, value];

      !cookieExist &&
        Cookie.set(key, JSON.stringify(mergeCookieValue), options);
    } else {
      Cookie.set(key, JSON.stringify([value]));
    }
  }

  static get<T>(key: string, defaultValue: T) {
    return Cookies.get(key) || defaultValue;
  }

  static getAll() {
    return Cookies.get();
  }

  static remove(key: string, options = {}) {
    return Cookies.remove(key, { ...Cookie.DEFAULT_OPTIONS, ...options });
  }
}

export default Cookie;
