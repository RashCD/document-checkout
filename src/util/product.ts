export type productDetailsTypes = {
  productID: string;
  productName: string;
  price: string;
  currency: '$';
  image: {
    src: string;
    alt: string;
  };
};

export type promoTypes = {
  code: string;
  product: string;
  discountPrice: number;
  description: string;
  eligibility: {
    discount: number;
    purchaseItem: number;
  };
};

export const PRODUCT_DETAILS: productDetailsTypes[] = [
  {
    productID: 'wf',
    productName: 'workflow',
    price: '199.99',
    currency: '$',
    image: {
      src: './workflow.jpg',
      alt: 'workflow image',
    },
  },
  {
    productID: 'docgen',
    productName: 'Document Generation',
    price: '9.99',
    currency: '$',
    image: {
      src: './docgen.jpg',
      alt: 'document generating image',
    },
  },
  {
    productID: 'form',
    productName: 'Form',
    price: '99.99',
    currency: '$',
    image: {
      src: './form.jpg',
      alt: 'form image',
    },
  },
  {
    productID: 'tax',
    productName: 'Taxation',
    price: '20.10',
    currency: '$',
    image: {
      src: './tax.jpg',
      alt: 'taxation image',
    },
  },
];

export const PRODUCT_PROMOCODE: promoTypes = {
  code: 'FF9543D1',
  product: 'docgen',
  discountPrice: 8.99,
  description:
    'Reduces the docgen price to $8.99 a unit when at least 10 documents are purchased',
  eligibility: {
    discount: 1.0,
    purchaseItem: 10,
  },
};
