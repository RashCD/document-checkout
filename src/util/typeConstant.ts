export const COOKIE_NAME = {
  PRODUCT_IDS: 'productIds',
};
