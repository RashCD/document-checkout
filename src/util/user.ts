export const PREFERENCES = {
  country: 'US',
  language: 'en-US',
  currency: '$',
};
