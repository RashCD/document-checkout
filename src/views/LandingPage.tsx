import { RouteComponentProps } from '@reach/router';

const LandingPage = (props: RouteComponentProps) => {
  var a = 'test';

  return (
    <div className="App">
      <header className="App-header">
        <p>test {a}</p>
      </header>
    </div>
  );
};

export default LandingPage;
