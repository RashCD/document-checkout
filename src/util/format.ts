import { PREFERENCES } from './user';

export const priceString = new Intl.NumberFormat(PREFERENCES.language, {
  maximumFractionDigits: 2,
  minimumFractionDigits: 2,
});

export const priceNumberFormat = (price: number): number =>
  Math.round((price + Number.EPSILON) * 100) / 100;
