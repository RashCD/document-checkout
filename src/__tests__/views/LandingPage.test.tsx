import React from 'react';
import {
  cleanup,
  findByText,
  fireEvent,
  getAllByRole,
  render,
} from '@testing-library/react';
import LandingPage from '../../views/LandingPage';
import { PRODUCT_DETAILS } from '../../util/product';
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from '@reach/router';

afterEach(cleanup);

function renderWithRouter(
  ui: JSX.Element,
  { route = '/', history = createHistory(createMemorySource(route)) } = {}
) {
  return {
    ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
}

describe('Landing page component test', () => {
  beforeEach(() => {
    renderWithRouter(<LandingPage />);
  });

  test('should contains "Doc Shopper"', async () => {
    expect(await findByText(document.body, 'Doc Shopper')).toHaveTextContent(
      'Doc Shopper'
    );
  });

  test('should have cart icon visible', async () => {
    const container = document.body;

    const cartButton = getAllByRole(container, 'button').find(
      (element) => element.className === 'button button'
    );

    expect(cartButton).toBeValid();

    expect(cartButton).toBeVisible();
  });

  test('should match number of product list', () => {
    const container = document.body;

    const totalProductList = PRODUCT_DETAILS.length;

    const matchElement = container.querySelector('.productList');

    expect(matchElement?.childElementCount).toEqual(totalProductList);
  });

  test('should shows the correct info in item listing', () => {
    const container = document.body;

    const productElement = container.querySelector('.productContainer');

    expect(productElement).toHaveTextContent('workflow');

    expect(productElement).toHaveTextContent('$199.99');

    expect(productElement?.lastElementChild?.childElementCount).toBe(2);
  });

  test('should add to cart count when clicking add to cart button', () => {
    const container = document.body;

    const addToCartButton = container.querySelector('.productAction')
      ?.firstChild as HTMLButtonElement;

    fireEvent.click(addToCartButton);

    expect(window.location.pathname).toBe('/');
  });

  test('should navigate to checkout page when clicking the buy now button', () => {
    const container = document.body;

    const buyNowButton = container.querySelectorAll(
      '.cta'
    )[1] as HTMLButtonElement;

    fireEvent.click(buyNowButton);

    expect(window.location.pathname).toBe('/checkout');
  });

  test('should match the snapshot', () => {
    const container = document.body;

    expect(container).toMatchSnapshot();
  });
});
