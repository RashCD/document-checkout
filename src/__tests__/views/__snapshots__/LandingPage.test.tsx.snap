// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing page component test should match the snapshot 1`] = `
<body>
  <div>
    <div
      class="app"
    >
      <header
        class="appHeader"
      >
        <button
          class="button backButton"
          type="button"
        >
          <span
            class="iconContainer"
            data-icon="true"
            style="--icon-src: url(back.svg); --icon-color: var(--white); --icon-size: 30px;"
          />
          BACK
        </button>
        <h1
          class="title"
        >
          Doc Shopper
        </h1>
      </header>
      <main
        class="appMain"
      >
        <div
          class="productList"
        >
          <div
            class="productContainer"
          >
            <img
              alt="workflow image"
              class="productImage"
              src="./workflow.jpg"
            />
            <div
              class="productTitle"
            >
              <p>
                workflow
              </p>
            </div>
            <div
              class="productPricing"
            >
              <p>
                $
                199.99
              </p>
            </div>
            <div
              class="productAction"
            >
              <button
                class="button cta"
                type="button"
              >
                Add to Cart
              </button>
              <button
                class="button cta"
                type="button"
              >
                Buy Now
              </button>
            </div>
          </div>
          <div
            class="productContainer"
          >
            <img
              alt="document generating image"
              class="productImage"
              src="./docgen.jpg"
            />
            <div
              class="productTitle"
            >
              <p>
                Document Generation
              </p>
            </div>
            <div
              class="productPricing"
            >
              <p>
                $
                9.99
              </p>
            </div>
            <div
              class="productAction"
            >
              <button
                class="button cta"
                type="button"
              >
                Add to Cart
              </button>
              <button
                class="button cta"
                type="button"
              >
                Buy Now
              </button>
            </div>
          </div>
          <div
            class="productContainer"
          >
            <img
              alt="form image"
              class="productImage"
              src="./form.jpg"
            />
            <div
              class="productTitle"
            >
              <p>
                Form
              </p>
            </div>
            <div
              class="productPricing"
            >
              <p>
                $
                99.99
              </p>
            </div>
            <div
              class="productAction"
            >
              <button
                class="button cta"
                type="button"
              >
                Add to Cart
              </button>
              <button
                class="button cta"
                type="button"
              >
                Buy Now
              </button>
            </div>
          </div>
          <div
            class="productContainer"
          >
            <img
              alt="taxation image"
              class="productImage"
              src="./tax.jpg"
            />
            <div
              class="productTitle"
            >
              <p>
                Taxation
              </p>
            </div>
            <div
              class="productPricing"
            >
              <p>
                $
                20.10
              </p>
            </div>
            <div
              class="productAction"
            >
              <button
                class="button cta"
                type="button"
              >
                Add to Cart
              </button>
              <button
                class="button cta"
                type="button"
              >
                Buy Now
              </button>
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</body>
`;
